Scripting API documentation (#14725)
mdn/content: files/en-us/mozilla/add-ons/webextensions/api/scripting/updatecontentscripts/index.md

---
title: scripting.updateContentScripts()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/updateContentScripts
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Non-standard
  - Reference
  - WebExtensions
  - updateContentScripts
  - scripting
browser-compat: webextensions.api.scripting.updateContentScripts
---
{{AddonSidebar()}}

Updates registered content scripts. If there are errors during script parsing and file validation, or if the IDs specified do not exist, no scripts are updated.

> **Note:** This method is available in Manifest V3 or higher.

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

### Syntax

```js
let removing = browser.scripting.updateContentScripts(
  scripts,         // object
  callback         // function
)
```

### Parameters

- `scripts`

  - : An object with these properties:  

    - `script`
      - : {{WebExtAPIRef("scripting.RegisteredContentScript")}}. Details of a script to update.

>     - `persistAcrossSessions`

Reviewer: This flag is part of RegisteredContentScript.

The type should just be "an array of RegisteredContentScript", similar to how `registerContentScripts` is documented.

>       - : `boolean`. Whether the content script persists into future sessions. 

Reviewer 2: I think we have to declare this prop and set it to `false` in FF for now, until we have full support for `persistAcrossSessions` ([Bug 1751436](https://bugzilla.mozilla.org/show_bug.cgi?id=1751436)).

- `callback`{{optional_inline}} 
  - : `function`. Invoked upon completion of the request.

### Return value

Return an array of {{WebExtAPIRef("scripting.RegisteredContentScript")}}.

Returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) when the callback parameter is not specified. If any error occurs, the promise is rejected with an error message.

## Examples

> This example updates a script file called `script.js`:

Reviewer 2: I think it rather updates a _registered content script_ called `"a-script"`.

> This example updates a script file called `script.js`:

Reviewer: Let's make this example more meaningful:

1. Register a script first.
2. Update the script by changing matches from example.com to `["https://example.com/*", "https://example.net/*"]` to show that one can change the set of URLs where the script is injected.

---
title: scripting.updateContentScripts()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/updateContentScripts
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Reference
  - WebExtensions
  - updateContentScripts
  - scripting
browser-compat: webextensions.api.scripting.updateContentScripts
---
{{AddonSidebar()}}

> Updates registered content scripts. If there are errors during script parsing and file validation, or if the IDs specified do not exist, no scripts are updated.

Reviewer 2: ```suggestion
Updates already registered content scripts. If there are errors during script parsing and file validation, or if the IDs specified do not exist, no scripts are updated.
```

> **Note:** This method is available in Manifest V3 or higher.

> This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

Reviewer 2: ```suggestion
 To use this API you must have the `"scripting"` [permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) and permission for the page's URL, either explicitly as a [host permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or using the [activeTab permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#activetab_permission).

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).
```

### Syntax

```js
await browser.scripting.updateContentScripts(
  scripts         // object
)
```

### Parameters

- `scripts`
  - : `array` of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. Details of a script to update. All the properties are optional except for `id`.

### Return value

A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. If any error occurs, the promise is rejected.

## Examples

This example updates a registered content script with ID `a-script` by setting `allFrames` to `true`:

```js

> await browser.scripting.updateContentScripts([
>   {
>     id: "a-script",
>     allFrames: true,
>   },
> ]);

Reviewer 2: ```suggestion
await browser.scripting.registerContentScripts([
  {
    id: "a-script",
    js: ["script.js"],
    matches: ["*://example.org/*"],
  },
]);

// Update content script registered before to allow execution in all frames:
await browser.scripting.updateContentScripts([
  {
    id: "a-script",
    allFrames: true,
  },
]);
```

---
title: scripting.updateContentScripts()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/updateContentScripts
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Non-standard
  - Reference
  - WebExtensions
  - updateContentScripts
  - scripting
browser-compat: webextensions.api.scripting.updateContentScripts
---
{{AddonSidebar()}}

Updates registered content scripts. If there are errors during script parsing and file validation, or if the IDs specified do not exist, no scripts are updated.

> **Note:** This method is available in Manifest V3 or higher.

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

### Syntax

```js

> let removing = browser.scripting.updateContentScripts(
>   scripts         // object
> )

Reviewer 2: ```suggestion
await browser.scripting.updateContentScripts(
  scripts         // object
)
```

```

### Parameters

> - `scripts`

Reviewer 2: `scripts` is an array of `RegisteredContentScript`

Reviewer 2: For updates, all but the `id` properties are optional in `RegisteredContentScript`.

  - : An object with these properties:  

    - `persistAcrossSessions`
      - : `boolean`. Whether the content script persists into future sessions. 
    - `script`
      - : `array` of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. Details of a script to update.

### Return value

> A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. If any error occurs, the promise fulfills with an error message.

Reviewer 2: It's similar to another sentence I've seen previously and I am confused: why is this described in the "return value" section? `scripting.updateContentScripts()` returns a promise that either resolves or rejects but when it resolves, there is no return value.

## Examples

> This example updates a script file called `a-script.js`:

Reviewer 2: ```suggestion
This example updates a content script that was previously registered with ID `a-script` by setting `allFrames` to `true`:
```

```js

> const aScript = {
>   id: "a-script",
>   js: ["script.js"],
>   matches: ["<all_urls>"],
>   };
> 
> await browser.scripting.updateContentScripts([aScript]);

Reviewer 2: ```suggestion
await browser.scripting.updateContentScripts([
  {
    id: "a-script",
    allFrames: true,
  },
]);
```

> **Note:** This method is available in Manifest V3 or higher.

To use this API you must have the `"scripting"` [permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) and permission for the page's URL, either explicitly as a [host permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or using the [activeTab permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#activetab_permission).

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

### Syntax

```js
await browser.scripting.updateContentScripts(
  scripts         // object
)
```

### Parameters

- `scripts`
  - : `array` of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. Details of a script to update. All the properties are optional except for `id`.

### Return value

A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. If any error occurs, the promise is rejected.

## Examples

> This example updates a registered content script with ID `a-script` by setting `allFrames` to `true`:

Reviewer 2: ```suggestion
This example updates a content script registered with ID `a-script` by setting `allFrames` to `true`:
```

```js

> await browser.scripting.registerContentScripts([
>   {
>     id: "a-script",
>     js: ["script.js"],
>     matches: ["*://example.org/*"],
>   },
> ]);
> 
> // Update content script registered before to allow execution in all frames:
> await browser.scripting.updateContentScripts([
>   {
>     id: "a-script",
>     allFrames: true,
>   },
> ]);

Reviewer 2: ```suggestion
try {
  await browser.scripting.registerContentScripts([
    {
      id: "a-script",
      js: ["script.js"],
      matches: ["*://example.org/*"],
    },
  ]);

  // Update content script registered before to allow execution in all frames:
  await browser.scripting.updateContentScripts([
    {
      id: "a-script",
      allFrames: true,
    },
  ]);
} catch (err) {
  console.error(`failed to register or update content scripts: ${err}`);
}
```

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

### Syntax

```js
await browser.scripting.updateContentScripts(
  scripts         // object
)
```

### Parameters

- `scripts`
  - : `array` of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. Details of a script to update. All the properties are optional except for `id`.

### Return value

A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of {{WebExtAPIRef("scripting.RegisteredContentScript")}}. If any error occurs, the promise is rejected.

## Examples

This example updates a content script registered with ID `a-script` by setting `allFrames` to `true`:

```js
try {
  await browser.scripting.registerContentScripts([
    {
      id: "a-script",
      js: ["script.js"],
      matches: ["*://example.org/*"],
    },
  ]);

>   // Update content script registered before to allow execution in all frames:

Reviewer 2: ```suggestion
  // Update content script registered before to allow execution
  // in all frames:
```

(a bit more readable in the preview IMO)

Author: Updated
